uniswap-v2-twap/src/UniswapV2Twap.sol

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: MIT
+// WARNING! This smart contract and the associated zk-SNARK verifiers have not been audited.
+// DO NOT USE THIS CONTRACT FOR PRODUCTION
+pragma solidity 0.8.19;
+
+import {Ownable} from "openzeppelin-contracts/access/Ownable.sol";
+import {IAxiomV1Query} from "axiom-contracts/contracts/interfaces/IAxiomV1Query.sol";
+import "utils/RLPReader.sol";
+
+contract UniswapV2Twap is Ownable {
+    using RLPReader for RLPReader.RLPItem;
+    using RLPReader for RLPReader.Iterator;
+    using RLPReader for bytes;
+
+    address public axiomQueryAddress;
+    mapping(bytes28 => uint256) public twapPris;
+
+    event UniswapV2TwapProof(
+        address pairAddress,
+        uint32 startBlockNumber,
+        uint32 endBlockNumber,
+        uint256 twapPri
+    );
+    event UpdateAxiomAddress(address newAddress);
+
+    constructor(address _axiomQueryAddress) {
+        axiomQueryAddress = _axiomQueryAddress;
+    }
+
+    function updateAxiomQueryAddress(
+        address _axiomQueryAddress
+    ) external onlyOwner {
+        axiomQueryAddress = _axiomQueryAddress;
+        emit UpdateAxiomAddress(_axiomQueryAddress);
+    }
+
+    function _getTimestampFromBlock(
+        uint32 blockNumber,
+        bytes32 blockHash,
+        bytes memory rlpEncodedHeader
+    ) internal pure returns (uint256) {
+        require(keccak256(rlpEncodedHeader) == blockHash, "invalid blockhash");
+
+        RLPReader.RLPItem[] memory ls = rlpEncodedHeader.toRlpItem().toList();
+        require(blockNumber == ls[8].toUint(), "invalid block number");
+        uint256 timestamp = ls[11].toUint();
+        return timestamp;
+    }
+
+    function _currentCumulativePrice(
+        uint112 reserve0,
+        uint112 reserve1,
+        uint256 blockTimestamp,
+        uint32 blockTimestampLast,
+        uint priceCumulativeLast
+    ) internal pure returns (uint256) {
+        //overflow is desired
+        unchecked {
+            uint256 increment = uint(reserve1 / reserve0) *
+                (blockTimestamp - blockTimestampLast);
+            return increment + priceCumulativeLast;
+        }
+    }
+
+    function _unpackReserveValues(
+        uint256 slot
+    ) internal pure returns (uint112, uint112, uint32) {
+        uint112 reserve0 = uint112(slot >> 144);
+        uint112 reserve1 = uint112(slot >> 32);
+        uint32 blockTimestampLast = uint32(slot);
+        return (reserve0, reserve1, blockTimestampLast);
+    }
+
+    function calculateUniswapV2Twap(
```

Comment:
It would be helpful to include the `forge inspect [contract] storage --pretty` output that gave the slot numbers below